processing/src/main/java/org/apache/druid/query/aggregation/firstlast/FirstLastUtils.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.druid.segment.DimensionHandlerUtils;
+import org.apache.druid.segment.NilColumnValueSelector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+
+import javax.annotation.Nullable;
+
+public class FirstLastUtils
+{
+
+  /**
+   * Returns whether a given value selector *might* contain object assignable from pairClass (SerializablePairLong*).
+   */
+  public static boolean selectorNeedsFoldCheck(
+      final BaseObjectColumnValueSelector<?> valueSelector,
+      @Nullable final ColumnCapabilities valueSelectorCapabilities,
+      Class pairClass
+  )
+  {
+    if (valueSelectorCapabilities != null && !valueSelectorCapabilities.is(ValueType.COMPLEX)) {
+      // Known, non-complex type.
+      return false;
+    }
+
+    if (valueSelector instanceof NilColumnValueSelector) {
+      // Nil column, definitely no SerializablePairLongObject.
+      return false;
+    }
+
+    // Check if the selector class could possibly be of pairClass* (either a superclass or subclass).
+    final Class<?> clazz = valueSelector.classOfObject();
+    return clazz.isAssignableFrom(pairClass)
+           || pairClass.isAssignableFrom(clazz);
+  }
+
+  @Nullable
+  public static SerializablePairLongDouble readDoublePairFromVectorSelectors(
+      @Nullable boolean[] timeNullityVector,
+      long[] timeVector,
+      Object[] objectVector,
+      int index
+  )
+  {
+    final long time;
+    final Double value;
+
+    final Object object = objectVector[index];
+
+    if (object instanceof SerializablePairLongDouble) {
+      // We got a folded object, ignore timeSelector completely, the object has all the info it requires
+      final SerializablePairLongDouble pair = (SerializablePairLongDouble) object;
+      // if time == null, don't aggregate
+      if (pair.lhs == null) {
+        return null;
+      }
+      time = pair.lhs;
+      value = pair.rhs;
+    } else {
+      if (timeNullityVector != null && timeNullityVector[index]) {
+        // Donot aggregate pairs where time is unknown
+        return null;
+      }
+      time = timeVector[index];
+      value = DimensionHandlerUtils.convertObjectToDouble(object);
+    }
+    return new SerializablePairLongDouble(time, value);
+  }
+
+  @Nullable
+  public static SerializablePairLongFloat readFloatPairFromVectorSelectors(
+      @Nullable boolean[] timeNullityVector,
+      long[] timeVector,
+      Object[] objectVector,
+      int index
+  )
+  {
+    final long time;
+    final Float value;
+
+    final Object object = objectVector[index];
+
+    if (object instanceof SerializablePairLongFloat) {
+      // We got a folded object, ignore timeSelector completely, the object has all the info it requires
+      final SerializablePairLongFloat pair = (SerializablePairLongFloat) object;
+      // if time == null, don't aggregate
+      if (pair.lhs == null) {
+        return null;
+      }
+      time = pair.lhs;
+      value = pair.rhs;
```

Comment:
I wonder why the need to create a new `SerializablePairLongFloat` - and not return the existing instance if there is one?

Reply:
Returning the same pair for now. I was worried about the reuse in the original source, however the SerializablePair class is immutable, so that shouldn't matter.